quaterion/train/xbm/xbm_buffer.py

```diff
@@ -0,0 +1,62 @@
+from typing import Tuple
+
+import torch
+from torch import LongTensor, Tensor
+
+from quaterion.train.xbm.xbm_config import XbmConfig, XbmDevice
+
+
+class XbmBuffer:
+    """A buffer implementation to hold recent N embeddings and target values.
+
+    Inspired by https://github.com/msight-tech/research-xbm/blob/master/ret_benchmark/modeling/xbm.py
+
+    Args:
+        config: Config class to configure XBM settings.
+    """
+
+    def __init__(self, config: XbmConfig, embedding_size: int):
+        if config.device == XbmDevice.AUTO:
+            device = "cuda" if torch.cuda.is_available() else "cpu"
+        else:
+            device = config.device
+
+        self._cfg = config
+
+        self._embeddings = torch.zeros((self._cfg.buffer_size, embedding_size)).to(
+            device
+        )
+        self._targets = torch.zeros(self._cfg.buffer_size, dtype=torch.long).to(device)
+        self._pointer = 0
+        self._is_full = False
+
+    @property
+    def is_full(self) -> bool:
+        return self._is_full
+
+    def get(self) -> Tuple[Tensor, LongTensor]:
+        if self.is_full:
+            return self._embeddings, self._targets
+        else:
+            return self._embeddings[: self._pointer], self._targets[: self._pointer]
+
+    def queue(self, embeddings: Tensor, targets: LongTensor) -> None:
+        """Queue batch embeddings and targets in the buffer.
+
+        Args:
+            embeddings: Output embeddings in the batch.
+            targets: Target values in the batch.
+        """
+        batch_size = len(targets)
+
+        if self._pointer + batch_size > self._cfg.buffer_size:
+            self._embeddings[-batch_size:] = embeddings
+            self._targets[-batch_size:] = targets
```

Comment:
In the current situation we can replace a part of the (n-1)-th batch:

Let's assume our buffer size is 250 and batch size is 16. 
We put 15 batches (15*16=240 embeddings), and in the next step we want to add  one more batch (16 embeddings). We put them into last 16 cells of the buffer. Hence, we replace 6 embeddings (pos 234-240) from the (n-1)-th batch, but I think it would be better to replace those in the positions 0-6.

Or we can make user to set buffer size proportional to batch size to avoid this problem.

Reply:
good catch 👍 handled it.